### ISSUE TITLE
Information request: h264 i-frame conversion

### ISSUE THREAD

**Issue by roleoroleo** [emericg/MiniVideo#4]
Thank you for sharing your project.
I'm trying to use it to convert an h264 raw stream to a bitmap image.
The stream is a sequence of 3 nalu: SPS, PPS and IDR (0x67, 0x68 and 0x65).

The API sequence I'm trying is the following:
```
minivideo_open
minivideo_parse
minivideo_decode_frame
minivideo_destroy_frame
minivideo_close
```

Is it correct?
Could you help me please?

**Comment by emericg** [comment]
Hi!
Yes that seems about right.

For instance here is a quick exemple that decode first I Frame into a QImage (for a QML image thumbnailer)
https://gist.github.com/emericg/8dd5f2601c9ff90b357222d44f2e23be

Does mini_thumbnailer works on your stream ?
You can also build mini_anayzer and setting the #define THUMBNAILS_ENABLED 1 in tabcontainer.cpp.
I just checked, if you drag and drop an elementary stream it will be parsed and thumbnailed

![Capture d’écran de 2021-03-27 16-37-02](https://user-images.githubusercontent.com/691258/112725876-b1369c80-8f1a-11eb-881e-113ec9585258.png)

**Comment by roleoroleo** [comment]
Sorry for the delay but i didn't see your answer.

> Does mini_thumbnailer works on your stream ?

No, this is the output:
```
root@debian:/home/user/h264/MiniVideo/mini_thumbnailer/build# ./mini_thumbnailer -i ./seq.h264
mini_thumbnailer main()
* mini_thumbnailer version 5.0

main() arguments
* input: ./seq.h264
minivideo_print_infos()
* Library version '0.13-0'
* Library built with C++ '201103'
* Library built with GCC '6.3.0'
* Library built on 'Mar 24 2021, 17:48:05'
* This is a DEBUG build
minivideo_print_features()
* STB_IMAGE_WRITE support is ON
* EXTERNAL WEBP support is OFF
* EXTERNAL JPEG support is OFF
* EXTERNAL PNG support is OFF
* COLORS are ON
* DEBUG traces are ON
[E][ MAIN]{minivideo.cpp:132} TEST error
[W][ MAIN]{minivideo.cpp:133} TEST warning
minivideo_endianness()
* Host system is LITTLE_ENDIAN

Working...
[E][ BITS]{bitstream.cpp:261} <b> Unable to read from the input file!
[E][ BITS]{bitstream.cpp:261} <b> Unable to read from the input file!
[E][H.264]{h264.cpp:516} 2 errors happened when decoding SPS and PPS content!

mini_thumbnailer exited without errors. Have a nice day.
```

[seq.zip](https://github.com/emericg/MiniVideo/files/7012792/seq.zip)

**Comment by emericg** [comment]
Hi, thanks for providing a test stream it made it easy to find the problem.
The elementary stream parser is not very evolved, and not used much. There was a problem related to the fact that you had only 3 samples in your stream, but that is now fixed!

As you can see there is still a decoding bug, but that I won't fix today ^^

![Capture d’écran de 2021-08-19 12-51-53](https://user-images.githubusercontent.com/691258/130056820-f6879b5d-7128-469c-a4a1-79492c2be1ff.png)

**Comment by roleoroleo** [comment]
Thank you for your work.

**Comment by roleoroleo** [comment]
Tested the new commit.
Now mini_thumbnailer doesn't exit with "Unable to read from the input file!" but it doesn't return successfully.
It remains blocked on "working":
```
root@debian:/home/user/h264/MiniVideo/mini_thumbnailer/build# ./mini_thumbnailer -i ./seq.h264
mini_thumbnailer main()
* mini_thumbnailer version 5.0

main() arguments
* input: ./seq.h264
minivideo_print_infos()
* Library version '0.13-0'
* Library built with C++ '201103'
* Library built with GCC '6.3.0'
* Library built on 'Aug 19 2021, 13:57:19'
minivideo_print_features()
* STB_IMAGE_WRITE support is ON
* EXTERNAL WEBP support is OFF
* EXTERNAL JPEG support is OFF
* EXTERNAL PNG support is OFF
* COLORS are ON
* DEBUG traces are OFF
minivideo_endianness()
* Host system is LITTLE_ENDIAN

Working...
```
until os terminates it.

**Comment by emericg** [comment]
Ho well, I got to say mini_thumbnailer was in a pretty sorry state.
Looking at the code, I guess many of it's features haven't been working for a couple of years ^^

But I was feeling nostalgic so I rewrote the faulty parts, let me know how it goes!

**Comment by roleoroleo** [comment]
It's working.
As you already said there is a decoding bug, but it works.
